umix-1.0.2/doc/umix.1
.\" $Id: umix.1,v 1.3 2003/02/09 09:35:45 sakari Exp $
.\"
.\" Manual page for umix. See README for more information.
.\" Copyright (C) 2002 Sakari Lehtonen <sakari@ionstream.fi>
.\"
.\"
.TH "UMIX" 1 "08 Jun 2002" "" "Sound Utilities"
.SH "NAME"
umix \- adjust soundcard mixers
.SH "SYNOPSIS"
.B umix
.RI [ \|channel\|\ volume\| | \|left:right ]\c
.RB [ R | P ]\c
.RB [\| + | - \|]
.IR device \|]
.RB [\| \-r
.IR driver \|]
.RB [\| \-f
.IR configpath \|]
.RB [\| \-ghilqsv \|]
.SH "DESCRIPTION"
.B umix
is a program for adjusting soundcard mixers.
With it you can adjust your soundcard volumes, balances and
recording sources.
It can be used flexibly from the command-line,
or from an interactive user interface.

You can save your soundcard volumes to a file, and load them
afterwards, for example when when booting the machine.

Channel names are used to specify the volume you want to
adjust. You can check what channels are supported with the 
.B -q
option. Channel names can also be shortened if they have enough
distinctive letters to identify the channel, eg. 'volume' can be 
shortened to 'vol'. If channel name is specified as
.I all
, all available channels are set to the specified value.

For each
.I channel
specified,
.I volume
is the volume you want to set, with both left
and right specified as the same, or in
.IR left : right
format.

You can also specify optional
.BR + \ or \ -
after the
.I volume
for increasing or decreasing the volume for
.BR channel ,
with amount
.IR volume .
The optional
.BR R \ or \ P
at the end of the
.I channel
argument sets the
.I channel
as a current recording source or normal play, respectively.
You can use both upper- and lowercase letters.

You can slide volumes smoothly in steps from one point to another.
The format is following:
.IR startvol , endvol , steps , delay .
Here
.I startvol
is the volume you want to slide the volume from, and
.I endvol
is the volume you want to increase or decrease in steps to. 
Optional parameter
.I steps
specifies the number of steps between the two volumes, and
.I delay
the delay between two steps. Default values for
.I steps
is 20, and 20ms for
.IR delay .
The optional
.BR + \ or \ -
have no effect when sliding volumes.

The
.I delay
can be between 0 and 5 seconds, specified in milliseconds.
Maximum value for 
.I steps 
is the difference between the starting
and destination volume, so that the minimum raise or decrease in one 
step is always 1.

All volume settings are parsed the same, so you can use sliding in your
config file too, you just have to edit the config manually.
For example the line

pcm 0,100,25,50

in your config file would slide the volume of 'pcm' from 0 to 100
in 25 steps and with a 50 ms delay when loading settings.

Umix needs read and write permissions to the mixer device. I suggest
that you create a group for this and change permissions
so that group has write and read access to the devices, then
add necessary users to that group.

.SH "OPTIONS"
.TP
.B \-q, \-\-query
Query and print all active channel volumes and stats from
all available devices. Interactive UI is not started.
.TP
.B \-l, \-\-load
Load previously saved settings from
.IR configpath .
If
.I configpath
is specified as '-', settings are loaded from
standard input. Clears the
.BR \-i
option.
.TP
.B \-s, \-\-save
Save current mixer settings to
.IR configpath .
If
.I configpath
is specified as '-', settings are printed to
standard output. Clears the
.BR \-i
option.
.TP
.BI \-f\  configpath ,\ \-\-file= configpath
Specify the file used to read and save settings.
Defaults to
.IR /etc/umixrc .
Sets the
.BR \-i
option.
.TP
.BI \-d\  device ,\ \-\-device= device
Specify the device used. Defaults to
.IR /dev/mixer .
Sets the
.BR \-i
option.
.TP
.BI \-r\  driver ,\ \-\-driver= driver
Specify the driver used. Defaults to
.IR OSS .
Sets the
.BR \-i
option.
.TP
.B \-i, \-\-interactive
Start interactive user interface after processing
non-interactive command-line options. 
.TP
.B \-g, \-\-globals
Print current global options. This is useful for finding out what
is the default device etc.
.TP
.B \-h, \-\-help
Display help and exit.
.TP
.B \-v, \-\-version
Display version information and exit.
.SH "EXAMPLES"
.TP
.B umix\ \-q
Prints all volumes of active channels and their stats.
The output should be self\-explanatory.
.TP
.BI umix\  volume\ 100 \  bass\ 88 \  treble\ 80 \  pcm\ 94
Sets the volume to 100, bass to 88,
treble to 80 and PCM to 94.
.TP
.BI umix\  vol\ 100 : 80 \  syn\ 90 : 30
Here channel names are shortened and stereo settings are used.
Sets volume`s left to 100, right to 80 and synth`s
left to 90 and right to 30.
.TP
.BI umix\  bass\ 20- \  pcm\ 20 : 10+ \  ogain\ 40 : 45-
Here volumes are incremented or decremented from their
original values. Decreases bass by 20, increases
PCM`s left by 20 and right by 10 and the last one
decreases output gain`s left by 40 and right by 45.
.TP
.BI umix\  mic\ R \  synth\ 20P \  line\ 10p-
This sets the mic as a recording source,
synth as normal play with volume 20 and line
as normal play with volume decreased by 10.
.TP
.BI umix\  pcm\ 0,60,30,25 \  vol\ 100,70 \  synth\ 50:50,25:50
Here sliding is demonstrated. The first slides pcm from 0 to 60
with 30 steps and a 25 ms delay, then slides vol from 100 to 70 with
default steps and delay and the last one slides synth from 50:50 to
25:50. 
.TP
.BI umix\ \-f \ /some/path/umixrc \ \-l
Here the config path is specified as
.I /some/path/umixrc
and settings are loaded from that file.
.TP
.BI umix\ \-d \ /dev/mixer2 \ \-f \ /somepath/umsettings
The device that is used for mixing is specified as
.IR /dev/mixer2 ,
config path is set as
.I /somepath/umsettings
and the interactive user interface is automatically
started, because.
.BR \-f \ and \ \-d
set the
.BR \-i
option.
.SH "NCURSES USER INTERFACE"
The (n)curses based user interface. Designed to be used
from a console or x-terminal.
.SS "USAGE"
On the left side is the channel menu, which consists of
channel names, stats and volumes.
On the right side is the balance menu.
The currently chosen channel is indicated with highlighting
its name and drawing arrows around the chosen menu.
You can move around with default
.BR vi / emacs / arrow \ keys.

Moving up and down selects the current channel, moving
left and right adjusts the volume or balance,
depending on which menu you are in.
.B Tab
toggles between volume and balance menus.

If the channel does not support stereo mixing
(it`s left and right values cannot differ), then a
.BR '|'
at center of the balance is drawn.
.SS "KEY BINDINGS"
.TP
.BR j \ / \ ^n \ / \ arrow\ down
Move down.
.TP
.BR k \ / \ ^p \ / \ arrow\ up
Move up.
.TP
.BR h \ / \ ^b \ / \ arrow\ left
Decrease volume / balance left.
.TP
.BR l \ / \ ^f \ / \ arrow\ right
Increase volume / balance right.
.TP
.BR ] \ / \ .
Increase volume by fifth of maximum amount.
.TP
.BR [ \ / \ ,
Decrease volume by fifth of maximum amount.
.TP
.B Tab
Toggle between volume and balance menus.
.TP
.B Space
Toggle play / record source.
.TP
.B L
Load settings.
.TP
.B S
Save settings.
.TP
.B f
Specify the config path. This presents a dialog in the lower
border. If no string is given, the path remains unchanged.
.TP
.B e
Enter channel volume manually for the current channel. 
Same rules apply as with command-line channel settings, 
so you can set balances and recording sources too.
.TP
.B n
Cycles through available mixer devices. You can see the currently
active mixer and the total number of initialized mixers in the
top border.
.TP
.B m
Mutes all those channels that are not already mute. If pressed
again, returns volumes back to the state before muting.
.TP
.B x
Excludes the current channel. Same as
.B m
but only with the current channel.
.TP
.B H
Displays the help.
.TP
.B q
Exit the program.
.TP
.B ^l
Refresh the screen.
.SH "CONFIG FILE"
The config consists of global options and mixer settings.
Global options come after the keyword
.BR [global] .
Global options are read before loading any mixer settings or
parsing command-line options. Global options are the defaults
to use if no command-line options are given.

Mixer settings come after the keyword
.BR [mixers] .
The mixer device that settings are loaded for is specified with the keyword
.BR mixer .
The string after
.BR mixer=
is the path of the mixer device. Then comes the settings
for that mixer. These are parsed like command-line options.

Comments and all other characters are ignored.

.SH "FILES"
.TP
.I /etc/umixrc
The default configuration file.
.TP
.I $HOME/.umixrc
Used as the config file if exists.
.SH "BUGS"
When used with terminals that have columns below 35, the output
is not optimal with the ncurses UI.

If the number of channels is greater than number of lines on screen, you
cannot adjust channels that exceed the number of lines in the ncurses UI 
(rare, but to be fixed soon)
.SH "AUTHOR"
Sakari Lehtonen <sakari@ionstream.fi>


umix-1.0.2/NEWS
Version 1.0.2
=============
Fixed a bug that might have caused segfaulting when quitting, and fixed a bug
that was introduced in 1.0.1 that caused the ncurses UI to display incorrect
values when loading or saving settings.

Version 1.0.1
=============
Minor bugfixes and updates. Updated Debian packages to build with correct
version number. You know what they say about final releases..

Version 1.0
============
Final version released. Removed plugin code, and unnecessary UIs. This is just
a simplified commandline and ncurses UI release, as I originally intended. 
I'm not working on Umix anymore, and this was a test project to begin with.

Version 0.9.3
=============
Umix has moved to SourceForge.Net! The new homepage URL is 
http://umix.sf.net/, check it out now! You can now also access the CVS tree
via anonymous access. Wow, CVS is the best thing since hmm .. RCS :) !

Version 0.9.2
=============
Minor bugfixes and tweaking. Default device with Linux is now /dev/sound/mixer,
you can use `configure --with-mixer-dev=path' to configure it.

Version 0.9.1
=============
GTK+2 user interface is now built by default. Note that the GTK+2 UI is still 
far from ready, but it is somehow usable by now. More verbose plugin handling 
and of course bugfixes and general polish.

Version 0.9.0
=============
Now includes an experimental GTK+2 user interface! See ChangeLog for details. 
Rewrote the API again, better plugin handling and general code cleanup.

Version 0.8.3
=============
Minor changes to error reporting and plugins are now built independently.

Version 0.8.2
=============
Added support for sliding volumes in steps and fixed some bugs. Now with
Slackware packages too.

Version 0.8.1
=============
Code and documentation were polished, and some bugfixes were added.

Version 0.8.0
=============
Major rewrite of almost everything, now uses a function based API to access 
mixer interfaces. New dynamic way of parsing channel options: channels are no 
longer parsed as commandline options but as channel names. Renamed some 
commandline options and of course bugfixes.

Version 0.7.2
=============
This release includes a new command line option for printing current global 
settings, a new option to enter volumes manually in ui_ncurses, fixes for 
command line options -p, -s, and --ui to actually work, and some other minor 
bugfixes.

Version 0.7.1
=============
A new way of scaling stuff in the ncurses UI, a bugfix in loading settings, 
removal of the GTK UI, a new third mixer device in the dummy driver, and some 
code cleanups.

Version 0.7.0
=============
All optional drivers and UIs are now dynamically loaded at runtime. 
Configuration file format rewritten so that global options and multiple mixer 
settings can be saved and loaded.

Version 0.6.0
=============
A rewrite for the whole mixer interface, including support for multiple mixer i
devices in one session, maybe even with different drivers.

Version 0.5
=============
Better error handling and various updates.

Version 0.4
=============
Internationalization support and some small fixes.

Version 0.3
=============
Initial public announcement.


umix-1.0.2/src/umix.c
/*
 * $Id: umix.c,v 1.9 2003/08/23 12:54:09 sakari Exp $
 *
 * The main function and command-line UI for Umix.
 * Main parses command-line options, initializes the mixer 
 * and then calls the UI. The command-line UI does not
 * support multiple devices in one session.
 *
 * Copyright (C) 2002 Sakari Lehtonen <sakari@ionstream.fi>
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 */

#include "umix_common.h"
#include "ui.h"
#ifdef HAVE_GETOPT_LONG
#  include <getopt.h>
#else
#  include <unistd.h>
#endif

/* global options */
umix_cfg globalopts =
{
	DEF_DEVICE_PATH,
	DEF_CONFIG_PATH,
	DEF_DRIVER
};

/* function prototypes */
static void check_homeconf(void);
static int cmdline(int argc, char *argv[], int optind);
static void cmdline_loadsettings(void);
static void parse_cmdlineopts(int argc, char *argv[]);
static void parse_cmdlinevols(int argc, char *argv[], int optind);
static void query(void);
static void printlevels(void);
static void usage(int status);
static void version(void);
static void showglobals(void);
static int setall(const char *optarg);

#ifdef UMIX_DEBUG
static int test(int mode);
#endif

/* the name umix is called with */
const char *program_name;

/* command line options */
typedef struct umix_opts umix_opts;
struct umix_opts
{
	int query;
	int load;
	int save;
	int force_interactive;
	int interactive;
} static opts = { 0, 0, 0, 0, 0 };

int main(int argc, char *argv[])
{
#ifdef ENABLE_NLS
	/* internationalization */
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
#endif
	program_name = argv[0];

	/* check if .umixrc exists in the users home dir, and set to the
	 * global config path if exists */
	check_homeconf();
	/* try to load global settings if not loading or saving to stdin */
	if (globalopts.cfgpath[0] != '-' && globalopts.cfgpath[1] != '\0')
		settings_load(globalopts.cfgpath, UMIX_SET_GLOBAL);

	/* parse command-line options */
	parse_cmdlineopts(argc, argv);
	
	/* check if umix was called with only one argument eg `umix foo` */
	if (argc == 2 && optind == 1)
	{
		err_msg(_("Nothing to do!"));
		usage(EXIT_FAILURE);
	}
	/* if any command line options given, parse them */
	if (argc >= 2)
	{
		if (cmdline(argc, argv, optind) == -1)
			return -1;
	}
	/* if no command line options were given, or -i was given, then start
	 * interactive UI, mixers are initialized in the UI */
	if (argc < 2 || opts.interactive == 1) 
	{
		/* no sense in saving or loading settings 
		 * from stdin/stdout in interactive UI's, replace
		 * config path with the default one */
		if (globalopts.cfgpath[0]=='-' && 
			globalopts.cfgpath[1]=='\0')
		{
			strncpy(globalopts.cfgpath, DEF_CONFIG_PATH,
				sizeof(globalopts.cfgpath));
		}
		if (ui_init("ncurses", argc, argv) == -1)
		{
			err_msg(_("Ncurses UI initialization failed"
			      ", try `%s -q'"), program_name);
		}
	}
	/* finally, close all mixers */
	close_mixers();

	return 0;
}

static void check_homeconf(void)
{
	char *str;
	struct stat sbuf;
	char buf[BUFLEN];

	/* check if $HOME/.umixrc exists, and use as the default
	 * config if found */
	if ((str = getenv("HOME")) != NULL)
	{
		snprintf(buf, sizeof(buf), "%s/.umixrc", str);
		if (stat(buf, &sbuf) == 0)
		{
			strncpy(globalopts.cfgpath, buf, 
				sizeof(globalopts.cfgpath));
		}
	}
}

/* the commandline UI */
static int cmdline(int argc, char *argv[], int optind)
{
#ifdef UMIX_DEBUG
	err_msg("cmdline: argc=%d optind=%d", argc, optind);
#endif
	if (init_mixers() == -1)
	{
		err_msg(_("Failed initializing mixer devices"));
		return -1;
	}

	/* process all command-line options conserning channels */
	parse_cmdlinevols(argc, argv, optind);
	/* if any channel options are given, do not start interactive ui */
	if (optind != argc)
		opts.interactive = 0;
	
	/* query settings */
	if (opts.query == 1)
	{
		query();
		/* no point in both querying levels and starting 
		 * the interactive ui */
		opts.interactive = 0;
	}
	
	/* load settings */
	if (opts.load == 1)
		cmdline_loadsettings();

	/* save settings */
	if (opts.save == 1)
	{
		if (settings_save(globalopts.cfgpath) == -1)
			err_str(_("Failed saving settings to \'%s\'"), 
				globalopts.cfgpath);
		else
			printf(_("Saved mixer settings to \'%s\'\n"), 
				globalopts.cfgpath);
		opts.interactive = 0;
	}
	/* -i forces to start the interactive ui after all other
	 * command line options are parsed */
	if (opts.force_interactive == 1)
		opts.interactive = 1; 

	return 0;
}

static void cmdline_loadsettings(void)
{
	int retval;

	retval = settings_load(globalopts.cfgpath, UMIX_SET_MIXER);
	if (retval == 0)
	{
		printf(_("Loaded mixer settings from \'%s\'\n"),
			globalopts.cfgpath);
	}
	else
	{
		err_msg("Loading settings failed");
		switch (retval)
		{
		case UMIX_ESETINVALID: 
			err_msg("%s: %s", 
				globalopts.cfgpath,
				UMIX_UI_EINVALID_STR); 
			break;
		case UMIX_ESETNOTFOUND:
			err_msg("%s: %s", 
				globalopts.cfgpath,
				UMIX_UI_ENOTFOUND_STR);
			break;
		case UMIX_ESETBINARY:
			err_msg("%s: %s", 
				globalopts.cfgpath,
				UMIX_UI_ESETBINARY_STR);
			break;
		default: 
			err_str("%s", globalopts.cfgpath);
			break;
		}
	}
	opts.interactive = 0;
}

static void parse_cmdlineopts(int argc, char *argv[])
{
	int optch; 
	const char *optstr = "d:r:f:ghlLqsSvi";
	
#ifdef HAVE_GETOPT_LONG
        const struct option longopts[] =
        {
		{ "device",	1, 0, 'd' },
		{ "driver",	1, 0, 'r' },	
                { "file",       1, 0, 'f' },
		{ "globals",	0, 0, 'g' },
                { "help",       0, 0, 'h' },
                { "load",       0, 0, 'l' },
                { "query",      0, 0, 'q' },
                { "save",       0, 0, 's' },
                { "version",    0, 0, 'v' },
		{ "interactive",0, 0, 'i' },
                { NULL,         0, 0, 0 }
        };
	while ((optch = getopt_long(argc, argv, optstr, longopts, NULL)) != -1)
	{
#else
	while ((optch = getopt(argc, argv, optstr)) != -1)
        {
#endif /* HAVE_GETOPT_LONG */
		switch(optch)
		{
		case 'd': strncpy(globalopts.devpath, optarg,
				  sizeof(globalopts.devpath));
			  opts.interactive = 1;
			  break;
		case 'f': strncpy(globalopts.cfgpath, optarg,
				  sizeof(globalopts.cfgpath));
			  opts.interactive = 1;
			  break;
		/* showglobals() is called here because we want to be
		 * able to call it even if something fails */
		case 'g': showglobals(); 
			  opts.interactive = 0;
			  break;
		case 'h': usage(EXIT_SUCCESS);
			  break;
		case 'i': opts.force_interactive = 1; 
			  break;
		case 'L': /* compatibility */
		case 'l': opts.load = 1; 
			  break;
		case 'q': opts.query = 1;
			  break;
		case 'r': strncpy(globalopts.driver, optarg,
				  sizeof(globalopts.driver));
			  opts.interactive = 1;
			  break;
		case 'S': /* compatibility */
		case 's': opts.save = 1;
			  break;
		case 'v': version(); 
			  exit(EXIT_SUCCESS);
			  break;
		default:
			  usage(EXIT_FAILURE);
			  break;
		}
        };
}

static void parse_cmdlinevols(int argc, char *argv[], int optind)
{
	int chnum;
	int retval;
	int i;
	
	for (i=optind; i<argc; i++)
	{
		/* the special channel name 'all' sets all channels */
		if ((strcasecmp(argv[i], "all")) == 0)
		{
			i++;
			setall(argv[i]);
		}
		/* get the right channel number for the string argv[i] eg.
		 * "pcm" would return 1 with my system */
		else if ((chnum = mixer_opt_to_chan_num(argv[i])) != -1)
		{
			i++;
			chan_set_curr(chnum);
			/* check if no value is given after the channel
			 * name */
			if (argv[i] == NULL)
			{
				err_msg(_("Value missing for channel %s"),
					chan_get_name());
				break;
			}
			/* parse the actual volume settings for current chan */
			retval = parsechanopt(argv[i]);
			if (retval != 0)
				err_msg(UMIX_EOPTGENERAL_STRFMT, 
					chan_get_name(), argv[i]);
			switch (retval)
			{
			case UMIX_EOPTLIMIT:
				err_msg(UMIX_EOPTLIMIT_STR);
				break;
			case UMIX_EOPTINVALID:
				err_msg(UMIX_EOPTINVALID_STR);
				break;
			case UMIX_EOPTNOTREC:
				err_msg(UMIX_EOPTNOTREC_STR);
				break;
			default:
				break;
			}
		}
		else
			err_msg(_("Invalid channel name: \'%s\'"), argv[i]);
	}
}

static void query(void)
{
	int i;

	for (i=0; i<mixer_get_num_mix(); i++)
	{
		mixer_set_curr(i);
		printlevels();
	}
}

/* parse optarg for all channels */
static int setall(const char *optarg)
{
        int i;

        for (i=0; i<mixer_get_num_chan(); i++)
        {
		chan_set_curr(i);
                if (parsechanopt(optarg) != 0)
			return -1;
        }

        return 0;
}

/* prints the levels and status for all channels on current mixer */
static void printlevels(void)
{
        int i;

	printf("%s::%s::%s\n", mixer_get_name(), 
		mixer_get_path(),
		mixer_get_driver());
        for (i=0; i<mixer_get_num_chan(); i++)
        {
		chan_set_curr(i);
		printf("%-10s %3d:%3d ", chan_get_name(), 
			chan_get_lr(CHAN_LEFT), chan_get_lr(CHAN_RIGHT));

		if (chan_is_stereo())
			printf(_("Stereo"));
		else
			printf(_("Mono  "));
		putchar(' ');

		if (chan_is_recsrc())
			printf(_("Record Source"));
		else if (chan_is_record())
			printf(_("Recordable"));
		putchar('\n');
        }
}

static void showglobals(void)
{
	printf(_("Current global settings:\n"
	       	"Config path\t   %s\n"
		"Device path\t   %s\n"
	       	"Mixer driver\t   %s\n"),
	       	globalopts.cfgpath, globalopts.devpath, 
		globalopts.driver);
}

static void usage(int status)
{
	if (status != 0)
		fprintf(stderr, _("try `%s -h' for more information\n"),
			program_name);
	else
	{
        printf(_("Usage: %s [channame volume|left:right][R|P][+|-] [options]\n"
	     "\nYou can get all available channel names with the '-q' option"
	     "\n\nAdditional options:\n"
             "-q, --query\t\tPrint channel settings and available\n"
	     "\t\t\tchannel names\n"
	     "-s, --save\t\tSave mixer settings to a file\n"
             "-l, --load\t\tLoad mixer settings from a file\n"
             "-f, --file=PATH\t\tSpecify file used to read and save settings\n"
	     "\t\t\t(defaults to %s)\n"
             "-d, --dev=PATH\t\tSpecify device used for mixing\n"
	     "\t\t\t(defaults to %s)\n"
	     "-r, --driver=DRIVER\tSpecify the mixer driver used "
	     "(defaults to %s)\n"
	     "-i, --interactive"
	     "\tStart interactive user interface after processing\n"
	     "\t\t\tcommand-line options (if available)\n"
	     "-g, --globals\t\tPrint current global options\n"
             "-h, --help\t\tDisplay this help and exit\n"
             "-v, --version\t\tDisplay version information and exit\n"),
		program_name, DEF_CONFIG_PATH, DEF_DEVICE_PATH, DEF_DRIVER);
	}
	exit(status);
}

static void version(void)
{
        printf(_("%s\n"
	"Copyright (C) 2002 Sakari Lehtonen <sakari@ionstream.fi>\n"
	"\nThis is free software; see the source for copying conditions. " 
	"There is NO\nwarranty; not even for MERCHANTABILITY or FITNESS FOR A" 
	" PARTICULAR PURPOSE.\n"), UMIX_VERSION_STR);
}

#ifdef UMIX_DEBUG
/* tests most of the functions available, and shows how to do things */
static int test(int mode)
{
	int i;
	int j;
	volume orig_vol;

	if (init_mixers() == -1)
	{
		err_msg(_("Failed initializing mixer devices"));
		return -1;
	}
	
	/* update check test */
	if (mode == 1)
	{
		while (1)
		{
			usleep(250000);
			if (mixer_get_status() == 1)
			{
				err_msg("mixer has been updated!");
				printlevels();
				chan_set_curr(0);
			}
		}

		mixer_close();
		return 0;
	}

	printf("Total number of mixers = %d\n", mixer_get_num_mix());
	for (j=0; j<mixer_get_num_mix(); j++)
	{
		printf("Mixer #%d information\n", mixer_get_curr());
		printf("numchan = %d\n", mixer_get_num_chan());
		printf("name    = %s\n", mixer_get_name());
		printf("path    = %s\n", mixer_get_path());
		printf("driver  = %s\n", mixer_get_driver());

		printf("num name       label      lt  rt   s r a m\n");
		for (i=0; i<mixer_get_num_chan(); i++)
		{
			chan_set_curr(i);
			printf("%3d %-10s %-10s ", chan_get_curr(), 
				chan_get_name(), chan_get_label());
			printf("%3d %3d  %d %d %d %d\n", 
				chan_get_lr(CHAN_LEFT), chan_get_lr(CHAN_RIGHT),
				chan_is_stereo(), chan_is_recsrc(), 
				chan_is_record(), chan_is_mute());
		}

		chan_set_curr(1);
		orig_vol = chan_get_vb();

		printf("Setting the channel %s to 100\n", chan_get_name());
		chan_set_lr(100, 100);
		usleep(200000);

		printf("Setting balance to right\n");
		chan_set_lr(0, 100);
		usleep(200000);

		printf("Setting balance to left\n");
		chan_set_lr(100, 0);
		usleep(200000);

		printf("Setting balance back to center\n");
		chan_set_lr(100, 100);

		printf("Decreasing volume with 2\n");
		chan_inc_vol(-2);
		err_msg("left=%d right=%d", chan_get_lr(CHAN_LEFT),
			chan_get_lr(CHAN_RIGHT));
		usleep(200000);

		printf("Increasing volume with 2\n");
		chan_inc_vol(2);
		err_msg("left=%d right=%d", chan_get_lr(CHAN_LEFT),
			chan_get_lr(CHAN_RIGHT));
		usleep(200000);

		printf("Restoring old values: vol=%3f bal=%3f\n", 
			orig_vol.volume, orig_vol.balance);
		chan_set_vb(orig_vol);

		for (i=mixer_get_num_chan()-1; i>=0; i--)
		{
			chan_set_curr(i);
			if (chan_is_record())
			{
				printf("Setting channel #%d as recsrc\n", i);
				chan_set_rec(CHAN_RECSRC);
				if (chan_is_recsrc())
				{
					printf("Channel #%d (%s) is recsrc\n",
						i, chan_get_name());
				}
				else
					printf("Setting recsrc failed!\n");

				break;
			}
		}
	}

	query();
	mixer_close();

	return 0;
}
#endif
